File: whatabook/ravenclaw_whatabook.py
""" 
    Title: ravenclaw_whatabook.py
    Author: Hannah Del Real and Tiffany Reyes
    Date: 14 July 2023
    Description: Connecting Python with What-a-Book database on MongoDB
"""

# Import mongoDBClient
from pymongo import MongoClient

# Create a connection string to connect to
client = MongoClient("mongodb+srv://web335_user:s3cret@bellevueuniversity.ozktyyu.mongodb.net/web335DBretryWrites=true&w=majority")

# Assign web335 database to variable db
db = client['web335DB']

# Create space between outputs
print("\n")


# Call the find function to display all books in the collection
def display_books():
    for book in db.books.find({}):
        print(f"{book['title']} by {book['author']} \n Details: \n  Genre: {book['genre']} \n  bookId: {book['bookId']} \n")

print("---Displaying all books in What-A-Book's collection--- \n")
print(display_books())


print("---See books by any of the following genres---")
genres = db.books.distinct("genre")
for genre in genres:
    print(genre)
    
print()


for genre in genres:
    print(f"Displaying books by the genre: {genre} \n")
    books = db.books.find({"genre": genre})
    for book in books:
        print(f"{book['title']} by {book['author']} \n Details: \n  Genre: {book['genre']} \n  bookId: {book['bookId']} \n")

File: whatabook/ravenclaw-whatabook.js
/*
===========================================================================================
; Title: ravenclaw-whatabook.js
; Author: Tiffany Reyes and Hannah Del Real
; Date: 11 July 2023
; Description: WhatABook queries
;==========================================================================================
*/

// Delete the books, customers, and wishlists collections.
db.books.drop();
db.customers.drop();
db.wishlists.drop();

// Create the books, customers, and wishlists collections.
db.createCollection("books", {
	validator: { $jsonSchema: {
		bsonType: "object",
        required: ["title", "author", "genre", "bookId"],
		properties: {
			title: {
				bsonType: "string"
			},
			author: {
				bsonType: "string"
			},
            genre: {
				bsonType: "string"
			},
			bookId: {
				bsonType: "string"
			}
		}
	}}
});

db.createCollection("customers", {
	validator: { $jsonSchema: {
		bsonType: "object",
        required: ["customerId", "firstName", "lastName"],
		properties: {
			customerId: {
				bsonType: "string"
			},
			firstName: {
				bsonType: "string"
			},
            lastName: {
				bsonType: "string"
			}
		}
	}}
});

db.createCollection("wishlists", {
	validator: { $jsonSchema: {
		bsonType: "object",
        required: ["wishlistId", "customerId", "wishlistbooks"],
		properties: {
            wishlistId: {
				bsonType: "string"
			},
            customerId: {
				bsonType: "string"
			},
			wishlistbooks: {
				bsonType: "array"
			}
		}
	}}
});

// Books
let book1 = {
    "title": "Pride and Prejudice",
    "author": "Jane Austin",
    "genre": "Romance",
    "bookId": "R001"
};

let book2 = {
    "title": "Harry Potter and the Prisoner of Azakaban",
    "author": "JK Rowling",
    "genre": "Fantasy",
    "bookId": "F003"
};

let book3 = {
    "title": "Harry Potter and the Goblet of Fire",
    "author": "JK Rowling",
    "genre": "Fantasy",
    "bookId": "F004"
};

// Insert book documents
db.books.insertOne(book1);
db.books.insertOne(book2);
db.books.insertOne(book3);

// Customers
let customer1 = {
    "customerId": "0003",
    "firstName": "Mina",
    "lastName": "Myoi"
};

let customer2 = {
    "customerId": "0004",
    "firstName": "Jihyo",
    "lastName": "Park"
};

// Insert customer documents
db.customers.insertOne(customer1);
db.customers.insertOne(customer2);

// Wishlists
let wishlist1 = {
    "wishlistId": "10001",
    "customerId": "0004",
    "wishlistbooks": [
        {
            "bookId": "F003"
        },
        {
            "bookId": "F004"
        }
    ]
};

let wishlist2 = {
    "wishlistId": "10000",
    "customerId": "0003",
    "wishlistbooks": [
        {
            "bookId": "R001"
        },
        {
            "bookId": "F003"
        }
    ]
};


// Insert wishlists documents
db.wishlists.insertOne(wishlist1);
db.wishlists.insertOne(wishlist2);

// Query to display list of books
db.books.find();

// Query to display list of books by genre
db.books.find( { "genre": "Fantasy" } );

// Query to display a list of books by author
db.books.find( { "author": "Jane Austin" } );

// Query to display a book by bookId
db.books.find( { "bookId": "R001" } );
